singularity-eos/eos/get_sg_eos_functors.hpp

```diff
@@ -216,12 +277,144 @@ struct final_functor {
     if (do_s) {
       sie_v(i) /= mass_sum;
     }
+    check_all_vals(i, npte, tid, pte_converged);
     /* reset mass fractions to original values if not normalized to 1 */
     for (int m = 0; m < nmat; ++m) {
       frac_mass_v(i, m) *= mass_sum;
     }
     return;
   }
+
+ private:
+  // Check to make sure returned values are normal or zero
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int bad_val(valT const value,
+                                            const char *const var_name) const {
+    const bool good =
+        (value == valT{0} || (std::isnormal(1e10 * value) && std::isnormal(value)));
+    if (not good) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
+             value);
+    }
+    return not good;
+  }
+  // Check for prrecluding zero and negative (non-positive) values
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int non_positive_val(valT const value,
+                                                     const char *const var_name) const {
+    const bool not_positive = value <= valT{0};
+    if (not_positive) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
+             value);
+    }
+    return not_positive;
+  }
+  // Check for prrecluding negative values
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int negative_val(valT const value,
+                                                 const char *const var_name) const {
+    const bool negative = value < valT{0};
+    if (negative) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
```

Comment:
Not to hammer on this point too much, but reducing code duplication is a Good Thing. All these "check" functions can be simplified:
```c++
// let's start using c++17! std::string_view and all
// if we roll a constexpr `isnormal` function, we replace the lambda with a constexpr eval
template<class Cond, class VT>
PORTABLE_FORCEINLINE_FUNCTION auto check_val (Cond&& C, VT&& value, std::string_view vname) const
{
  auto pass = Cond(std::forward<VT>(value));
  if (Cond(std::forward(value))) {
    PortsOfCall::printf("%s %e\n", var_name.data(), value);
  }
  return pass;
}

auto negative_val = [](auto value, auto name) { return check_val ([](auto v) { return v < 0; }, name); };
// and so on;
```
(spitballing a bit, probably a cleaner approach).

Reply:
I'll roll those into a single function. I just got lazy and added those negative checks at the end.

I'm also still in C++ 14 mode so I haven't started using string views.

Reply:
I don't think string views are necessary but I agree with @mauneyc-LANL that these checks should be unified and maybe moved to shared scope.

Reply:
I generalized this into the error utilities header